dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,206 @@ public Pattern getParameterPattern() {
-                allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+            try (ResultSet rs = getColumns()) {
+                while (rs.next()) {
+                    String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+                    allParams.put(NameNormaliser.normaliseName(paramName),
+                                  createColumnAccessor(paramName, rs));
+                }
             }
-            rs.close();
             return allParams;
         }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, getParamDirection(rs));
+        }
+
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            Direction paramDir = columnTypeToDirection(rs);
+            if (paramDir == null) {
+                throw new SQLException("Invalid parameter type");
+            }
+            return paramDir;
+        }
+
+        abstract protected Direction columnTypeToDirection(ResultSet rs) throws SQLException;
+
+        protected DbParameterAccessor createDbParamAccessor(String paramName, ResultSet rs, Direction direction)
+                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            return createDbParameterAccessor(paramName, direction,
+                           typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                           getJavaClass(paramTypeName), rs.getInt("ORDINAL_POSITION") - 1);
+        }
+    }
+
+    private class DatabaseTable extends DatabaseObject {
+
+        private DatabaseTable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        protected Direction columnTypeToDirection(ResultSet rs) {
+            return Direction.INPUT;
+        }
+    }
+
+    abstract private class DatabaseExecutable extends DatabaseObject {
+
+        private DatabaseExecutable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, getParamDirection(rs));
+        }
+
+        protected Direction columnTypeToDirection(ResultSet rs) throws SQLException {
+            return jdbcTypeToDirection(rs.getInt("COLUMN_TYPE"));
+        }
+
+        abstract protected Direction jdbcTypeToDirection(int jdbcType);
```

Comment:
Would it work if we use the same `columnTypeToDirection` name here? Also I'm not quite sure about the name of `jdbcType`.